### PR TITLE
Synchronize local folder with dropbox

### PR DESCRIPTION
**Pull request by robert00s**
Option to synchronize files in local koreader folder with folder on Dropbox. All files that aren't on the local disk will be downloaded. Files with different sizes will also be downloaded and overwritten.
The download process can be paused or stopped at any time (similar to downloading images in Wikipedia).

Synchronize and settings are available after long press in dropbox account in Cloud storage. 
Limitations: Only one folder (without subfolders) can be synchronize with one folder on local storage.

Close: #4971

### CODE REVIEW

File: frontend/apps/cloudstorage/cloudstorage.lua [hunk 1]
```diff
@@ -255,45 +292,254 @@ function CloudStorage:cloudFile(item, path)
...
+    Trapper:wrap(function()
+        Trapper:setPausedText("Download paused.\nDo you want to continue or abort downloading files?")
+        local ok, downloaded_files, failed_files = pcall(self.downloadListFiles, self, item)
+        if ok and downloaded_files then
+            if not failed_files then failed_files = 0 end
+            local text
+            if downloaded_files == 0 and failed_files == 0 then
+                text = _("No files to download from dropbox.")
+            elseif downloaded_files > 0 and failed_files == 0 then
+                text = T(_("Successfuly downloaded %1 files from dropbox to local disk."), downloaded_files)
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/cloudstorage.lua]:
```suggestion
                text = T(_("Successfuly downloaded %1 files from Dropbox to local storage."), downloaded_files)
```

File: frontend/apps/cloudstorage/cloudstorage.lua [hunk 2]
```diff
@@ -255,45 +292,254 @@ function CloudStorage:cloudFile(item, path)
...
+            else
+                text = T(_("Successfuly downloaded %1 files from dropbox to local disk.\nFailed downloaded %2 files."),
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/cloudstorage.lua]:
```suggestion
                text = T(_("Successfuly downloaded %1 files from Dropbox to local storage.\nFailed downloaded %2 files."),
```

File: frontend/apps/cloudstorage/cloudstorage.lua [hunk 3]
```diff
@@ -255,45 +292,254 @@ function CloudStorage:cloudFile(item, path)
...
+                    downloaded_files, failed_files)
+            end
+            UIManager:show(InfoMessage:new{
+                text = text,
+                timeout = 3,
+            })
+        else
+            Trapper:reset() -- close any last widget not cleaned if error
+            UIManager:show(InfoMessage:new{
+                text = _("No files to download from dropbox.\nCheck your configuration and connection."),
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/cloudstorage.lua]:
```suggestion
                text = _("No files to download from Dropbox.\nPlease check your configuration and connection."),
```

File: frontend/apps/cloudstorage/cloudstorage.lua [hunk 4]
```diff
@@ -255,45 +292,254 @@ function CloudStorage:cloudFile(item, path)
...
+        end
+    end
+
+    if files_to_download == 0 then
+        UI:clear()
+        return 0
+    end
+
+    local response, go_on
+    local procceded_file = 0
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/cloudstorage.lua]: Typo for processed?

Comment (robert00s) [reply frontend/apps/cloudstorage/cloudstorage.lua]: Yes, thanks :)

**Comment by Frenzie** [comment]
No one else commented, so I'll merge this now.

File: frontend/apps/cloudstorage/cloudstorage.lua [hunk 5]
```diff
@@ -255,45 +292,254 @@ function CloudStorage:cloudFile(item, path)
...
+function CloudStorage:synchronizeCloud(item)
+    local Trapper = require("ui/trapper")
+    Trapper:wrap(function()
+        Trapper:setPausedText("Download paused.\nDo you want to continue or abort downloading files?")
+        local ok, downloaded_files, failed_files = pcall(self.downloadListFiles, self, item)
+        if ok and downloaded_files then
+            if not failed_files then failed_files = 0 end
+            local text
+            if downloaded_files == 0 and failed_files == 0 then
+                text = _("No files to download from dropbox.")
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/cloudstorage.lua]: Whoops, missed this.

File: frontend/apps/cloudstorage/cloudstorage.lua [hunk 6]
```diff
@@ -255,45 +292,254 @@ function CloudStorage:cloudFile(item, path)
...
+            elseif downloaded_files > 0 and failed_files == 0 then
+                text = T(_("Successfuly downloaded %1 files from Dropbox to local storage."), downloaded_files)
+            else
+                text = T(_("Successfuly downloaded %1 files from Dropbox to local storage.\nFailed downloaded %2 files."),
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/cloudstorage.lua]: And this

File: frontend/apps/cloudstorage/cloudstorage.lua [hunk 7]
```diff
@@ -255,45 +292,254 @@ function CloudStorage:cloudFile(item, path)
...
+    local syn_dialog
+    local dropbox_sync_folder = item.sync_source_folder or "not set"
+    local local_sync_folder = item.sync_dest_folder or "not set"
+    syn_dialog = ButtonDialogTitle:new {
+        title = T(_("Dropbox folder:\n%1\nLocal folder:\n%2"), dropbox_sync_folder, local_sync_folder),
+        title_align = "center",
+        buttons = {
+            {
+                {
+                    text = _("Choose dropbox folder"),
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/cloudstorage.lua]: And this :-P

**Pull request merged**